src/parsers/gene_association_file.py

```diff
@@ -0,0 +1,301 @@
+"""
+GO Gene Association File (GAF) parser.
+
+This script processes Gene Ontology (GO) annotation data from GAF-style CSV files and normalizes it into a structured format compatible with ontology-driven data models.
+
+Main Functionality:
+--------------------
+- Loads raw GO annotations from an input CSV file.
+- Drops unused or redundant columns to simplify the dataset.
+- Constructs a unified 'subject' identifier by combining 'DB' and 'DB_Object_ID'.
+- Renames selected columns to match a target schema (e.g., predicate, object, supporting_objects).
+- Converts annotation dates into standard YYYY-MM-DD format.
+- Adds metadata fields such as 'aggregator' and 'protocol_id'.
+- Determines predicate negation based on a whitelist of allowed GO relations.
+- Maps evidence codes to ECO terms by merging with an external mapping file.
+- Applies fallback evidence types when no direct evidence mapping is found.
+- Outputs a clean, standardized CSV file for downstream processing.
+
+Resources:
+----------
+- Gene Association File (GAF) format specification: https://geneontology.org/docs/go-annotation-file-gaf-format-2.2/
+- Downloadable GAF files: http://current.geneontology.org/annotations/index.html
+- ECO (Evidence and Conclusion Ontology) documentation: https://www.evidenceontology.org/annotation_resources/#eco_go_mappings
+- GAF-to-ECO mapping file: http://purl.obolibrary.org/obo/eco/gaf-eco-mapping.txt
+
+Usage:
+------
+Run via CLI:
+    python -m parsers.gene_association_file --input annotations_data100.csv --output normalized_annotation.csv
+
+"""
+
+import pandas as pd
+import click
+import os 
+import sys
+
+# --- Constants ---
+
+SUBJECT = "subject"
+PREDICATE = "predicate"
+OBJECT = "object"
+PUBLICATIONS = "publications"
+EVIDENCE_CODE = "Evidence_Code"
+SUPPORTING_OBJECTS = "supporting_objects"
+ANNOTATION_DATE = "annotation_date"
+PRIMARY_KNOWLEDGE_SOURCE = "primary_knowledge_source"
+AGGREGATOR = "aggregator"
+PROTOCOL_ID = "protocol_id"
+NEGATED = "negated"
+EVIDENCE_TYPE = "evidence_type"
+
+ASSOCIATION_COL_HEADERS = [
+    SUBJECT,
+    PREDICATE,
+    OBJECT,
+    PUBLICATIONS,
+    EVIDENCE_CODE,
+    SUPPORTING_OBJECTS,
+    ANNOTATION_DATE,
+    PRIMARY_KNOWLEDGE_SOURCE,
+    AGGREGATOR,
+    PROTOCOL_ID,
+    NEGATED,
+    EVIDENCE_TYPE,
+]
+
+ASSOCIATION_COL_TYPES = {
+    SUBJECT: str,
+    PREDICATE: str,
+    OBJECT: str,
+    PUBLICATIONS: str,
+    SUPPORTING_OBJECTS: str,
+    ANNOTATION_DATE: str,
+    PRIMARY_KNOWLEDGE_SOURCE: str,
+    AGGREGATOR: str,
+    PROTOCOL_ID: str,
+    EVIDENCE_TYPE: str,
+    NEGATED: bool,
+}
```

Comment:
The columns don't need to be in a specific order, so you could either (a) use `list(ASSOCIATION_COL_TYPES.keys())` to set the value of `ASSOCIATION_COL_HEADERS` or (b) just use `ASSOCIATION_COL_TYPES.keys()` instead of setting the `ASSOCIATION_COL_HEADERS` variable.

Note that after python 3.7, dictionary keys are guaranteed to be in insertion order -- i.e. they will be in the same order as you have written them in the code, so you don't really need a separate list. You can also use `OrderedDict` from `collections`.